java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,179 @@
+package arcs.core.analysis
+
+import arcs.core.data.expression.Expression
+
+/** Field [Identifier]. Lists of [Identifier]s imply an AccessPath.*/
+typealias Identifier = String
+
+/**
+ * Result of the [ExpressionClaimDeducer] on Paxel [Expression]s.
+ *
+ * [Deduction]s recursively associate identifiers to field paths. These relationships can be
+ * directly translated into [Claim]s and [AccessPath]s.
+ */
+sealed class Deduction {
+
+  /** Unwrap [Deduction] object to get an underlying [Path]. */
+  abstract fun getPath(): Path
+
+  /** Substitute aliases in all paths. */
+  abstract fun substitute(aliases: Scope): Deduction
+
+  /** Union of two [Deduction] collections. */
+  abstract operator fun plus(other: Deduction): Deduction
+
+  /** [Deduction]s that have a simple path (a list of [Identifier]s). */
+  interface Pathlike {
+    val path: List<Identifier>
+    fun mergeTop(other: Pathlike): Deduction
+  }
+
+  /** A representation of a field path in a Paxel [Expression]. */
+  data class Path(override val path: List<Identifier> = emptyList()) : Deduction(), Pathlike {
+
+    constructor(vararg paths: Identifier) : this(paths.toList())
+
+    /** Base-case: Returns self as underlying path.*/
+    override fun getPath(): Path = this
+
+    /** Return [Path] with all alias substitutions applied. */
+    override fun substitute(aliases: Scope): Path =
+      Path(
+        path.flatMap { identifier ->
+          aliases.associations.getOrDefault(identifier, Paths(Path(identifier))).getPath().path
+        }
+      )
+
+    /** Append more [Identifier]s to the [Path] */
+    override fun mergeTop(other: Pathlike) = Path(path + other.path)
+
+    /** Union of a [Path] and a [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Path -> Paths(this) + Paths(other)
+      is Paths -> Paths(this) + other
+      is Scope -> Paths(this) + Paths(other)
+      else -> this + other
+    }
+  }
+
+  /** A collection of field [Path]s. */
+  data class Paths(val paths: List<Deduction> = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Deduction) : this(paths.toList())
+
+    constructor(vararg paths: List<Identifier>) : this(paths.map { Path(it) })
+
+    /** Union of a [Paths] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Path -> this + Paths(other)
+      is Paths -> Paths(paths + other.paths)
+      is Scope -> this + Paths(other)
+      else -> this + other
+    }
+
+    /** Returns the first [Path] in the collection, or an empty [Path]. */
+    override fun getPath(): Path = if (paths.isNotEmpty()) paths.first().getPath() else Path()
```

Comment:
How can you have getPath() on all Deductions if some deductions are not paths? Having a method that looks valid but internally throws away all paths but the first one is dangerous.